Added timeouts to wait for transmission loops

manutremo opened the pull request:
Hi, I added timeouts to wait for transmission loops to avoid the sketch getting stuck if the memory chip is not found of the bus fails for some reason, i.e. no pullups, broken wire, etc.

I used the existing MAX_ITER constant but hits might be changed depending on how long we need to wait.

As the timeout action I copied the action on the default: case of the corresponding switch statement. Please review if this is correct since I'm not a great specialist on how the I2C bus works.

I hope this works!

manutremo committed:
Added timeouts to wait for transmission loops

Hi, I added timeouts to wait for transmission loops to avoid the sketch getting stuck if the memory chip is not found of the bus fails for some reason, i.e. no pullups, broken wire, etc.

I used the existing MAX_ITER constant but hits might be changed depending on how long we need to wait.

As the timeout action I copied the action on the default: case of the corresponding switch statement. Please review if this is correct since I'm not a great specialist on how the I2C bus works.

I hope this works!

manutremo committed:
Fixed timeout condition

kisse66 commented:
thanks, I'll have a look.

kisse66 commented:
Did this fix your case?

I'll build a test setup (no HW at hand, all in use...), but there's two worries looking at the code

1) the timeout for waiting TWINT might be too short. Looping 200 times to check one bit does not take long. It may be fine, but I'd need to check so it don't randomly abort when everything should be ok
2)  begin:  does not have counter?  There's many jumps to there also

by the way it's actually quite obvious now why this code may freeze without pullups - HW will wait forever if the line stays low when trying to start. Your fix should already handle that case.

kisse66 commented:
..and perhaps the 
`if (m == MAX_ITER)`
will not catch the timeout since the loop exits with MAX_ITER+1  ?

I'll soon have a setup I can measure a bit.

manutremo committed:
Add separate timeout + fixed while exit condition

Hi, thanks for having a look.

Following on your last comments, added a separate timeout so it is possible to have different wait times. Set to 5000 just by intuition, feel free to modify according to your experience.

Also fixed the while exit condition.

Regarding the counter for begin:... honestly no idea if that would be necessary, i'll let it to your own judgement if including it or not :)

I should also have an available board in some days so I should be able to run some tests myself.

manutremo committed:
Tests done

I run some tests with a Mini Pro 8Mhz and I2C memory breadboard.
I updated the timeout detection and added error codes so that which timeout was triggered can be determined.
If everything works ok, the serial device will output SSF2XA. If the first timeout is triggered, then the output will be SSF21XA; if it was the second timeout, it will be SSF22XA, and so on.

I run the bootloader after disconnecting the wires from the memory and it seems to work correctly. I upped MAX_TIMEOUT to 250; and the I2C bus seems to respond within that period. should it be necessary to increase it, I guess the m variable should be upgraded to uint16.

manutremo committed:
Error codes only printed if DEBUG_ON

Error codes only printed if DEBUG_ON

kisse66 commented:
Extra debug codes are good addition!
I also tested with 8MHz and looks good. Still the start-wait may be a bit short. I ran it in simulator and with 16MHz it takes around 140us to time out.
But as it seems to work better than before let's fix that later if there's need.
I'll make one more check with 16MHz and merge it. 
With debug it's now 1612 bytes.

kisse66 commented:
Yep, works with 16MHz also and tried 3 different EEPROMs.

kisse66 commented:
Do you want to squash these separate commits into one or just keep as is?

manutremo commented:
Happy to know that it's working. I'm not sure I understand the difference between the two options, so I don't have any preference and leave it to your own judgment :)
